exchanges/kucoin/kucoin_websocket.go

```diff
@@ -0,0 +1,1069 @@
+		StartTime:  resp.Candles.StartTime,
+		Interval:   intervalString,
+		OpenPrice:  resp.Candles.OpenPrice,
+		ClosePrice: resp.Candles.ClosePrice,
+		HighPrice:  resp.Candles.HighPrice,
+		LowPrice:   resp.Candles.LowPrice,
+		Volume:     resp.Candles.TransactionVolume,
+	}
+	return nil
+}
+
+func (ku *Kucoin) processOrderbook(respData []byte, subject, instrument string) error {
+	response := WsOrderbook{}
+	var err error
+	if subject == "level2" {
+		result := WsLevel2Orderbook{}
+		err = json.Unmarshal(respData, &result)
+		response.Symbol = instrument
+		response.Changes.Asks = result.Asks
+		response.Changes.Bids = result.Bids
+		response.TimeMS = result.TimeMS
+	} else {
+		err = json.Unmarshal(respData, &response)
+	}
+	if err != nil {
+		return err
+	}
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	base := orderbook.Base{
+		Exchange:        ku.Name,
+		VerifyOrderbook: ku.CanVerifyOrderbook,
+		LastUpdated:     time.UnixMilli(response.TimeMS),
+		Pair:            pair,
+		Asset:           asset.Spot,
+	}
+	for x := range response.Changes.Asks {
+		item := orderbook.Item{}
+		item.Price, err = strconv.ParseFloat(response.Changes.Asks[x][0], 64)
+		if err != nil {
+			return err
+		}
+		item.Amount, err = strconv.ParseFloat(response.Changes.Asks[x][1], 64)
+		if err != nil {
+			return err
+		}
+		if response.Changes.Asks[x][2] != "" {
+			item.ID, err = strconv.ParseInt(response.Changes.Asks[x][2], 10, 64)
+			if err != nil {
+				return err
+			}
+		}
+		base.Asks = append(base.Asks, item)
+	}
+	for x := range response.Changes.Bids {
+		item := orderbook.Item{}
+		item.Price, err = strconv.ParseFloat(response.Changes.Bids[x][0], 64)
+		if err != nil {
+			return err
+		}
+		item.Amount, err = strconv.ParseFloat(response.Changes.Bids[x][1], 64)
+		if err != nil {
+			return err
+		}
+		if response.Changes.Bids[x][2] != "" {
+			item.ID, err = strconv.ParseInt(response.Changes.Bids[x][2], 10, 64)
+			if err != nil {
+				return err
+			}
+		}
+		base.Bids = append(base.Bids, item)
+	}
+	return ku.Websocket.Orderbook.LoadSnapshot(&base)
+}
+
+func (ku *Kucoin) processMarketSnapshot(respData []byte, instrument string) error {
+	response := WsSpotTicker{}
+	err := json.Unmarshal(respData, &(response))
+	if err != nil {
+		return err
+	}
+	pair, err := currency.NewPairFromString(instrument)
+	if err != nil {
+		return err
+	}
+	ku.Websocket.DataHandler <- &ticker.Price{
+		ExchangeName: ku.Name,
+		AssetType:    asset.Spot,
+		Last:         response.LastTradedPrice,
+		Pair:         pair,
+		Low:          response.Low,
+		High:         response.High,
+		QuoteVolume:  response.VolValue,
+		Volume:       response.Vol,
+		LastUpdated:  time.UnixMilli(response.Datetime),
+	}
+	return nil
+}
+
+// Subscribe sends a websocket message to receive data from the channel
+func (ku *Kucoin) Subscribe(subscriptions []stream.ChannelSubscription) error {
+	return ku.handleSubscriptions(subscriptions, "subscribe")
```

Comment:
I'm getting many of these:
`[ERROR] | WEBSOCKET |  13/01/2023 16:07:56  | Kucoin subscription failure:`